### ISSUE TITLE
Allow injecting resource ID refs into variables

### ISSUE THREAD

**Issue by dudicoco** [crossplane/crossplane#3385]
### What problem are you facing?

Currently, crossplane supports injecting resource ids into via `IDRefs`, for example:
```yaml
vpcSecurityGroupIDRefs:
- name: example
```

This is important in order to avoid hardcoding IDs.

However, there are some other cases where it is necessary to inject resource IDs into the configuration, for example, with IAM policies.
With some IAM policies you need to specify the resource ID:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::111122223333:oidc-provider/oidc.eks.region-code.amazonaws.com/id/EXAMPLED539D4633E53DE1B71EXAMPLE"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringEquals": {
                    "oidc.eks.region-code.amazonaws.com/id/EXAMPLED539D4633E53DE1B71EXAMPLE:sub": "system:serviceaccount:default:my-service-account",
                    "oidc.eks.region-code.amazonaws.com/id/EXAMPLED539D4633E53DE1B71EXAMPLE:aud": "sts.amazonaws.com"
                }
            }
        }
    ]
}
```
```json
{
   "Version": "2012-10-17",
   "Statement": [
      {
         "Effect": "Allow",
         "Action": [
             "rds-db:connect"
         ],
         "Resource": [
             "arn:aws:rds-db:us-east-2:1234567890:dbuser:db-ABCDEFGHIJKL01234/db_user"
         ]
      }
   ]
}
```

### How could Crossplane help solve your problem?

Allow injecting resources IDs into the configuration, possibly as a variable so it would be easy to apply it to the policy:
```yaml
ResourceIDRefs:
- variableName: dbExampleResourceID
  apiVersion: rds.aws.crossplane.io/v1alpha1
  kind: DBInstance
  resourceName: example
```
```json
{
   "Version": "2012-10-17",
   "Statement": [
      {
         "Effect": "Allow",
         "Action": [
             "rds-db:connect"
         ],
         "Resource": [
             "arn:aws:rds-db:us-east-2:1234567890:dbuser:${dbExampleResourceID.status.dbiResourceID}/db_user"
         ]
      }
   ]
}
```

**Comment by bobh66** [comment]
It's pretty common in these scenarios to have to patch multiple different values into the target ARN, like the account ID, region and target reference ID (role or other identifier), so a CombineFromComposite is used to map all of the relevant values into the target field (assumeRolePolicy for example).   The ID ref is usually patched from the status field of the relevant resource into a placeholder in metadata.annotations and then patched into the target field.

**Comment by dudicoco** [comment]
@bobh66 I don't believe that would help in such cases.

Composition only works if you group your resources within the composition, however, if for example you create an EKS cluster and you need to set many different `sts:AssumeRoleWithWebIdentity` policies for different applications, you would not include these policies in the same composition, they would reside in different files/directories.

So basically the variables need to be able to be injected without including the associated resources within the same composition.
In addition, it would provide a better user experience if the user prefers to use just managed resources without compositions (such as in my case).

**Comment by ytsarev** [comment]
@dudicoco FWIW, you can nest XRs within another superset Composition and share the data between them through status. An https://github.com/upbound/platform-ref-aws/ based example can be found [here](https://github.com/upbound/platform-ref-aws/blob/main/package/cluster/composition.yaml)

**Comment by dudicoco** [comment]
@ytsarev can you elaborate on how that works exactly? From my understanding it's not possible use observe only resources yet: https://github.com/crossplane/crossplane/issues/1722

In addition, I still think that getting ID refs should be possible outside of compositions, is there an objection to such a feature?

**Comment by ytsarev** [comment]
@dudicoco we can publish arbitrary information from MR within the Composition to the XR status and then consume this information from XR status to another MR within the same Composition. It is not an observe-only but an arbitrary data exchange mechanism as of now.

**Comment by dudicoco** [comment]
@ytsarev can you please share an example that reference the EKS OIDC information within the IAM role trusted policy?

**Comment by ytsarev** [comment]
@dudicoco so we can grab oidc info with this patch https://github.com/upbound/platform-ref-aws/blob/main/package/cluster/eks/composition.yaml#L71-L73 to status and then consume it in any other MR like here https://github.com/upbound/platform-ref-aws/blob/main/package/cluster/eks/composition.yaml#L200-L202

**Comment by dudicoco** [comment]
@ytsarev the example you provided patches a string (`spec.forProvider.url`).

An IAM role trusted policy is a json/yaml document, how do you patch only the relevant fields within the json?

**Comment by ytsarev** [comment]
@dudicoco you can do string `transform` similar to https://github.com/upbound/uxp-training/blob/main/lab-xrd-status/configuration/composition.yaml#L229-L249.

**Comment by dudicoco** [comment]
Thanks @ytsarev.

How do I transform multiple fields?

**Comment by ytsarev** [comment]
@dudicoco https://docs.crossplane.io/v1.9/reference/composition/#hl-7-4 `CombineFromComposite` should help :)

**Comment by dudicoco** [comment]
Thanks again @ytsarev.
Is it possible to make fields conditional (`if else`)?

**Comment by ytsarev** [comment]
@dudicoco conditionals are not supported by Compositions API, the advanced logic will be a part of Composition Functions implementation, please see design doc https://github.com/crossplane/crossplane/pull/2886

**Comment by dudicoco** [comment]
@ytsarev which brings me again to the proposed feature of this issue - if crossplane allowed for general injection of variables we could use other tools to handle the rest of the logic.

For example, helm could easily handle the EKS OIDC variable substitution within the IAM trusted policy while allowing it to be conditional, the syntax is also more straight forward.

**Comment by bobh66** [comment]
I think patching from environments - https://github.com/crossplane/crossplane/pull/3007 - is close to what you are asking for.  It essentially allows you to specify one or more sets of context information that can be patched into the resources within the composition.  This is effectively variable substitution without requiring logic within Crossplane or new syntax.  And it doesn't require the use of external tools like helm or kustomize to provide the information required.

**Comment by dudicoco** [comment]
@bobh66 can you please elaborate on how that feature solves the OIDC patching into the IAM trusted role policy described above?
In addition, #3007 is for compositions, while what i'm suggesting here is to be able to use resource refs within standard resources, it seems that this issue is similar to what i'm requesting: https://github.com/crossplane/crossplane/issues/1770

**Comment by bobh66** [comment]
The idea is that when you create a Cluster you also create an EnvironmentConfig for that cluster that contains all of the Cluster-specific information you might need later, like the OIDC information.  Then you can reference that EnvironmentConfig dynamically by name in a separate Composition.

This is indeed based on using Compositions, as that's what gives Crossplane most of it's functionality and value.  Reading through #1770 it does not appear to have settled on an implementation yet, so it's impossible to say whether it would apply to standalone managed resources or would require Compositions.
